File: app/src/main/java/com/example/assignment/utils/DiffUtilOptionAdapter.java
package com.example.assignment.utils;

import androidx.recyclerview.widget.DiffUtil;

import com.example.assignment.fragments.OptionBottomSheetDialog.*;
import com.example.assignment.models.Category;

import java.util.ArrayList;
import java.util.Objects;

public class DiffUtilOptionAdapter extends DiffUtil.Callback {
    private ArrayList newList;
    private ArrayList oldList;
    private OptionBottomSheetType type;

    public DiffUtilOptionAdapter(ArrayList newList, ArrayList oldList, OptionBottomSheetType type) {
        this.newList = newList;
        this.oldList = oldList;
        this.type = type;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        if (type == OptionBottomSheetType.CATEGORY) {
            return Objects.equals(((Category) newList.get(newItemPosition)).getTitle(), ((Category) oldList.get(oldItemPosition)).getTitle());
        } else {
            return Objects.equals((String) newList.get(newItemPosition), (String) oldList.get(oldItemPosition));
        }
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        if (type == OptionBottomSheetType.CATEGORY) {
            return Objects.equals(((Category) newList.get(newItemPosition)).getTitle(), ((Category) oldList.get(oldItemPosition)).getTitle()) &&
                    Objects.equals(((Category) newList.get(newItemPosition)).getThumbnail(), ((Category) oldList.get(oldItemPosition)).getThumbnail()) &&
                    Objects.equals(((Category) newList.get(newItemPosition)).getId(), ((Category) oldList.get(oldItemPosition)).getId()) &&
                    Objects.equals(((Category) newList.get(newItemPosition)).getDisplayColor(), ((Category) oldList.get(oldItemPosition)).getDisplayColor());
        } else {
            return Objects.equals((String) newList.get(newItemPosition), (String) oldList.get(oldItemPosition));
        }
    }
}

File: app/src/main/java/com/example/assignment/utils/DiffUtilOptionAdapter.kt
package com.example.assignment.utils

import androidx.recyclerview.widget.DiffUtil
import com.example.assignment.uis.fragments.OptionBottomSheetDialog.OptionBottomSheetType
import com.example.assignment.models.Category

class DiffUtilOptionAdapter(
    private val newList: ArrayList<*>,
    private val oldList: ArrayList<*>,
    private val type: OptionBottomSheetType
) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return if (type === OptionBottomSheetType.CATEGORY) {
            (newList[newItemPosition] as Category).title == (oldList[oldItemPosition] as Category).title
        } else {
            newList[newItemPosition] as String == oldList[oldItemPosition] as String
        }
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return if (type === OptionBottomSheetType.CATEGORY) {
            (newList[newItemPosition] as Category).title == (oldList[oldItemPosition] as Category).title && (newList[newItemPosition] as Category).thumbnail == (oldList[oldItemPosition] as Category).thumbnail && (newList[newItemPosition] as Category).id == (oldList[oldItemPosition] as Category).id && (newList[newItemPosition] as Category).displayColor == (oldList[oldItemPosition] as Category).displayColor
        } else {
            newList[newItemPosition] as String == oldList[oldItemPosition] as String
        }
    }
}